### ISSUE TITLE
Lifecycles passed Event.data and specially event.data.params object is unreliable.

### ISSUE THREAD

**Issue by sazaam** [strapi/strapi#17096]
it threatens the whole Lifecycles method and approach, as it makes the code complicated, unnecessarily long and forces to manage if/else cases, that could and should be avoided

### Required System information

- Node.js version: --- 16.14.2
- yarn version: --- 1.22.10
- Strapi version:---  4.11.1
- Database: --- MySQL
- Operating system: --- Windows 10
- Is your project Javascript or Typescript: -- Javascript

### Describe the bug

The data contained in Event.data from Lifecycles is subject to serious variations and makes it unreliable data

To reproduce this:
* Create section Data-Type, with name, level and path and parent field(many to one-related)
* after we enter the name and the parent, the level and path field will be automatically filled by the lifecycle hook method called (at beforeCreate and beforeUpdate states)

Expected:
* To have safe ways to access data from the FORM we just ENTERED , completely related to the item we are currently editing!!!

### Notes
So to recap, the case scenario is that you make after creation / update treatments to your models cases: when a parent is specified, levels have to be fetched from the specified parent same, when a parent is specified, a path has to be created along the parent path + '/' + the child path. Nothing extreme here, that is why the first code straight-forwardness and readible etc... and did the job.

Now,  you have to use event.params.data that if I understand correctly have some info about the current update or create request (where:{%id%} is even there etc...) but this api is such a headache when trying to retrieve other data from the object that is created/updated currently edited right now.

let us say that parent is a relation field where it HAS ONLY ONE POSSIBLE PARENT : sometimes it has 
```
parent:{connect:[Object], disconnect:[]}
// whatever it is it is still an array, while nobody dis ask for it to be an array since this relation field is many to ONE. so this one should be one element, no reason for it to be an array.
// another overcomplicated thing is that now what is this connect / disconnect thing? is that really a good standard for naming relations? is that also that accurate?
```

The main impacting issue here is that the behavior of this (and the contents of the event.data.params) of these lifecycles has become  much more unreliable than it was, and now has three (at least) different scenarios that forces using nested weird if/elses and still does not seem to solve issue seamlessly. -> overcomplicated => very messy code and inaccurate solving

i.e : there are cases of a created item (here, section) that will have this event.data and parameters set to overly complicated stuff when addressing the relation fields (parent:{connect:[Object], disconnect:[]})
sometimes, while updating it would try to show that there are no connected objects but there are disconnected objects as such `(parent:{connect:[], disconnect:[Object]})`

Please inform us about the clarity and helpfullness in the code of such a feature. how do you easily check for these in JS without getting a bunch of extra + dirty looking lines of code... And the weirdest part is when just 'Publishing'... now publishing ALSO has a different method, and still triggers the beforeUpdate, but with absolutely no reliable/related/accurate data in the event.data object. This kind of behavior is too messy and inadequate for clean declaration in lifecycles. (which use to be very straightforward and seamless) now because the data of the forms being completely scattered, the incoming event param and its children have no longer correct information about the current edited object / form thus beacome heavily unreliable as it kind of contains what it wants with no consistency.

Don't we expect to be able to read at least the data from the current form ???

### Steps to reproduce the behavior

1. Go to 'sections'
2. Click on 'edit one' or create one
3. edit fields per-say 'level' int field and 'parent' relational field
4. See what kind of data is in the event data passed to the lifecycle methods declared.
Sometimes empty, sometimes has parent:(connected:[somearray],disconnected:[]) sometimes the opposite (connected:[],disconnected:[somearray]) and sometimes has a totally different data style like parent:{id:13} 
so this kind of behavior is absolutely nasty looking and something has to be done at least to try to uniform it or to make api not usable this way.

### Expected behavior

in event.data.params are the actual fields we are editing in their (ready-to-submit) state.
i.e name:thename, parent:theparentObj, level, null, path:null, since we only entered name and parent
but this should be the case on EVERY update + on publishing! why will it have another behavior on publishing.
It would also make sense to have the same behavior on created as well.
but right now updating the fields on strapi dashboard makes the event.data object so different and variating that it makes all treatments hacky and unreliable.

### Code snippets

This code used to work flawlessly in V3.0  #lifecycles declarations in api/section/models/section.js
```
module.exports = {
  lifecycles: {
    // Called before an entry is created
    async beforeCreate(data) {
      await strapi.services.section.setLevel(data) ;
      await strapi.services.section.setPath(data) ;
    },
    // Called before an entry is updated
    async beforeUpdate(data) {
      await strapi.services.section.setLevel(data) ;
      await strapi.services.section.setPath(data) ;
    }
  },
};
```

and the corresponding code  on api/section/services/section.js
```
module.exports = {
  async setLevel (data) {
    if(!data.parent){
      data.level = 1 ;
    }else{
      let parent = await strapi.api.section.services.section.findOne({id:data.parent}) ;
      data.level = parent.level + 1 ;
    }
    return data ;
  },
  async setPath (data) {
    if(!data.parent){
      data.path = '/' + data.name + '/' ;
    }else{
      let parent = await strapi.api.section.services.section.findOne({id:data.parent}) ;
      data.path = parent.path + data.name + '/' ;
    }
    return data ;
  }
}
```

absolutely transparent and self-describing, both functions doing similar treatments beforecreate/update.
But now, the api works with more layers of unuseful complexity 👎  

```
/// beforecreate  By the way even this code described does not suffice to handle all cases
async beforeCreate(event) {
const { data, where, select, populate } = event.params;
let par;
if(!!data.parent.connect.length){
	par = await strapi.entityService.findOne('api::section.section', data.parent.connect[0].id)	;
}
await strapi.service('api::section.section').setLevel(event.params.data, par);
await strapi.service('api::section.section').setPath(event.params.data, par);
}
/// beforeupdate
async beforeUpdate(event) {
const { data, where, select, populate } = event.params;
let par;
if(!!data.parent){
	
	if(!!data.parent.connect.length){
		par = await strapi.entityService.findOne('api::section.section', data.parent.connect[0].id)	;	
	}
	console.log("PLOUF", par)
	
	if(data.parent.id){
		
		par = data.parent;
	}
				
}else {
	let item = await strapi.entityService.findOne('api::section.section', where.id, {populate:["parent"]}) ;
	par = item.parent;
}
await strapi.service('api::section.section').setLevel(event.params.data, par);
await strapi.service('api::section.section').setPath(event.params.data, par);
}
```

and still this code is bugged, it is not stable, now such kinds of lifecycle functions come with such a destroyed api 👎 
Plus look at why such impacting changes should be neccessary in BeforeCreate and BeforeUpdate? While it was seamlessly working fine in earlier versions.

### Additional context

Even with the %relationfield%:{connected:[somearray],disconnected:[]} kind of API that I dislike but it should have upsides as well, why sometimes the even.data comes with relationfield:{id:13} (the actual item) and sometimes it just would come without it (mostly when it was unchanged) : %relationfield% simply absent, just createdAt and updatedAt kind of fields...
this whole event data object seems so randomly populated that it greatly sacrifices and questions its usefulness. We want an object containing our edited data from right now. where did this information disappeared?? and why is it that tricky to retrieve?

Sorry, Dear Strapi Team to enounce this as a *Bug* because it is probably not one, and maybe intended somehow, but still it can only appear as a handicapping feature, which does not help and oppositely draws down severely the overall ease of use of the lifecycle feature.

I still see no clever workaround this that will maintain my code readable, nice and neat. That change forces a different approach onto us developers, and most probably a less straightforward and unnecessarily complicated one.

**Comment by galaxxyz** [comment]
Hey! I'll leave this comment just to make the issue more visible and everyone who face it could vote for it on the Public Roadmap here, even though i'm not sure if implementing it on Entity Service level will solve it: https://feedback.strapi.io/developer-experience/p/add-lifecycles-at-a-higher-level-entityservice
Unfortunately, for cases when an app logic becomes more complex, it's often impossible to avoid using lifecycles, and it would be very helpful to have more reliable lifecycles API.

Here are some examples when I personally used to use them and which seem to me pretty common usecases:
1. Validation of entries on relational fields (its existence, number for 1-to-N relations, and other custom rules). Lifecycles are the only way to implement it to work for changes coming from any source: API requests, Admin or Entity Service.
2. Creating related entries in beforeCreate/beforeUpdate hooks (e.g. you create an order and want to create a shipping at the same time to connect them with relation)
3. Calling 3rd party as a side effect and store the returned data in the same create/update cycle (works with simple properties, doesn't work with relations and media)

Basically, the problem is beforeCreate and beforeUpdate hooks. In after____ hooks we don't need to modify current object anymore, however to read relations or media we still need to call findOne Entity Service.

Trying to upgrade Strapi to more recent version where this new structure with connect/disconnect and the absence of media fields access was introduced, I have found myself in a situation when it's not feasible to refactor some features, mostly because the resulting code is unstable and hard to support and debug.

I understand, that this may not be the thing you would like to prioritise for future developments, or maybe there were some technical reasons for this change, but by introducing new lifecycles API in v4.5.0 we have lost significant functionality for strapi customisation.

**Comment by sazaam** [comment]
@galaxxyz Hi !

I will try to recap all that.

in the admin,
Create <datatype-you-want> Data-Type, with 
  name:string, 
  level:uint, 
  path:string,
  parent:relation(many to one-related, unique parent) 

(the goal was to have smart auto filling methods for 'level' and 'path' fields in the lifecycle hooks at beforeCreate and beforeUpdate states).

still in the admin
Go to <datatype> to 'edit one' or create one
edit fields per-say 'name' string field and 'parent' relational field

(it should work by writing level from the parent's level and writing path from joined parent's path + child's path)

the admin part is done.

in the strapi lifecycle hooks
(the part where you create files in your tree to overwrite strapi default behavior)

api/<datatype>/models/<datatype>.js
api/<datatype>/services/<datatype>.js

(sorry it has been a while and I omitted this in the steps)
that used to be in V.3
new location in v.4 :
src/api/<datatype>/content-types/<datatype>/lifecycles.js

that should look like this :
```
module.exports = {
	async beforeCreate(event) {
		const { data, where, select, populate } = event.params;
		//
		console.log(data.parent)
	},
  
	async beforeUpdate(event) {
		const { data, where, select, populate } = event.params;
		// 
		console.log(data.parent)
	},
  };
```


As I am retrying today the {id:13} case does not really show up but instead undefined shows up

how to have that :

Publish any unpublished <datatype> that actually has a parent.

now the parent does not come as {disconnected:[], connect:[{id:13}]}

but is clearly undefined.


Let me explain how I see the {id:13} case by memory 
I usually had to make a separate findOne query on the parent since I wanted extra information on the parent.
but sometimes when logging data.parent it actually came with the parent itself (with full data)

in my code I was forced to use both methods, i.e
- having arrays of disconnected/connected identifiers then requesting these <datatype>.findOne(<identifier>)
- sometimes be ready to retrieve directly the right parent, not as arrays but one item, with his own id, etc...

today I am getting undefined,
which is basically a very similar issue, undefined or {id:13} means that ONE parent is not here, 
while what it should do, is giving me an array of empty disconnected and empty connected, not undefined ^^.

I tried many things on that project but the back-end stayed the same, it also has been some time,

but thank you a lot for your attention.

**Comment by pinkasey** [comment]
I came across this difficulty today.
I had chatGPT write a utility function for me - one that normalizes the fields to "simple" format (not the connect/disconnect/set format).

The problem is that this function needs to know if the field it is transforming is a x-to-one or x-to-many field- in both cases you'd get `connect: [...]`. You need to know if to take the array, or just the first item. For that you need access to the schema, which complicates this function a lot.
In my case I opted to just hard-code the few fields that I know I currently need.

In any case,
I would expect Strapi to provide such functionality. 
Something like `event.apply()` that returns the entity as it would've been after applying the event (create/update).
